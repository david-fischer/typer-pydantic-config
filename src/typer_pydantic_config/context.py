import click
from pydantic import BaseModel

CONTEXT_CONFIG_KEY = "context"


def _get_config(ctx: click.Context) -> BaseModel:
    return ctx.obj[CONTEXT_CONFIG_KEY]


def set_config(ctx: click.Context, config: BaseModel) -> None:
    ctx.ensure_object(dict)
    ctx.obj[CONTEXT_CONFIG_KEY] = config


def get_config() -> BaseModel:
    return _get_config(click.get_current_context())
