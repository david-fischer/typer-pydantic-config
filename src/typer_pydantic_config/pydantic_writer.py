from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any

import toml
from pydantic import BaseModel


class PydanticWriter[PydanticModel: BaseModel](ABC):
    path: Path
    pydantic_cls: type[PydanticModel]

    def __init__(self, path: Path, pydantic_cls: type[PydanticModel]) -> None:
        self.path = path
        self.pydantic_cls = pydantic_cls

    def delete(self) -> None:
        self.path.unlink()

    def exists(self) -> bool:
        """Return if path already exists."""
        return self.path.exists()

    def update_on_disk(self, **update: Any) -> PydanticModel:
        """Update and save to disk."""
        config = self.load()
        update = {key: value for key, value in update.items() if value}
        updated_config = config.model_copy(update=update)
        self.save(updated_config)
        return updated_config

    @abstractmethod
    def load(self) -> PydanticModel: ...

    @abstractmethod
    def save(self, config: PydanticModel) -> None: ...

    def get_str_repr(self) -> str:
        """Return string representation of config."""
        return f"# {self.path}\n{self.path.read_text()}"


class ConfigTomlWriter[PydanticModel: BaseModel](PydanticWriter[PydanticModel]):
    def load(self) -> PydanticModel:
        """Load config from TOML or return defaults if file does not exist."""
        data = toml.load(self.path)
        return self.pydantic_cls(**data)

    def save(self, config: PydanticModel) -> None:
        """Persist the config to TOML on disk."""
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self.path.open("w", encoding="utf-8") as f:
            toml.dump(config.model_dump(), f)
