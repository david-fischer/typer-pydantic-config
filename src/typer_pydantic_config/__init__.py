from .config_app import start_config_app
from .context import get_config

__version__ = "0.1.0"

__all__ = [
    "get_config",
    "start_config_app",
]
