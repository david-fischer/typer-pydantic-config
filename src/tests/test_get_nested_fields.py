from pydantic import BaseModel
from syrupy import SnapshotAssertion

from typer_pydantic_config.click_utils import (
    get_flat_fields,
    get_nested_fields,
    update_pydantic_model_command,
)


class Inner(BaseModel):
    value: int
    name: str


class NestedModel(BaseModel):
    some_bool: bool
    value: int
    inner: Inner


def test_nested_pydantic_model(snapshot: SnapshotAssertion) -> None:
    snapshot.assert_match(get_nested_fields(NestedModel))


def test_nested_pydantic_model_flattened(snapshot: SnapshotAssertion) -> None:
    snapshot.assert_match(get_flat_fields(NestedModel))


def test_nested_pydantic_model_flattened_(snapshot: SnapshotAssertion) -> None:
    snapshot.assert_match(
        update_pydantic_model_command(NestedModel, lambda _: None).params
    )
