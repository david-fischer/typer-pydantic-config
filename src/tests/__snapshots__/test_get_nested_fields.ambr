# serializer version: 1
# name: test_nested_pydantic_model
  dict({
    'inner': dict({
      'name': FieldInfo(annotation=str, required=True),
      'value': FieldInfo(annotation=int, required=True),
    }),
    'some_bool': FieldInfo(annotation=bool, required=True),
    'value': FieldInfo(annotation=int, required=True),
  })
# ---
# name: test_nested_pydantic_model_flattened
  dict({
    'inner.name': FieldInfo(annotation=str, required=True),
    'inner.value': FieldInfo(annotation=int, required=True),
    'some_bool': FieldInfo(annotation=bool, required=True),
    'value': FieldInfo(annotation=int, required=True),
  })
# ---
# name: test_nested_pydantic_model_flattened_
  list([
    <Option some_bool>,
    <Option value>,
    <Option inner__SEP__value>,
    <Option inner__SEP__name>,
  ])
# ---
